modules/imgproc/test/test_emd.cpp

```diff
@@ -1,93 +1,196 @@
+
+    // expected results
+    const double emd0 = 2460. / 210;
+    Matx<float, 4, 5> flow0 {0, 0, 50, 0, 0, 0, 0, 20, 0, 40, 30, 20, 0, 0, 0, 0, 0, 0, 30, 20};
+
+    // basic call with cost
     {
-        16, 16, 13, 22, 17,
-        14, 14, 13, 19, 15,
-        19, 19, 20, 23,  M,
-        M ,  0,  M,  0,  0
-    };
-    static float  w1[] = { 50, 60, 50, 50 },
-                  w2[] = { 30, 20, 70, 30, 60 };
-    Mat _w1(4, 1, CV_32F, w1);
-    Mat _w2(5, 1, CV_32F, w2);
-    Mat _cost(_w1.rows, _w2.rows, CV_32F, cost);
-
-    float emd = EMD( _w1, _w2, -1, _cost );
-    if( fabs( emd - emd0 ) > success_error_level*emd0 )
+        float emd = 0.f;
+        ASSERT_NO_THROW(emd = EMD(w1, w2, DIST_USER, cost));
+        EXPECT_NEAR(emd, emd0, 1e-6 * emd0);
+    }
+
+    // basic call with cost and flow output
     {
-        ts->printf( cvtest::TS::LOG,
-            "The computed distance is %.2f, while it should be %.2f\n", emd, emd0 );
-        code = cvtest::TS::FAIL_BAD_ACCURACY;
+        Mat flow;
+        float emd = 0.f;
+        ASSERT_NO_THROW(emd = EMD(w1, w2, DIST_USER, cost, nullptr, flow));
+        EXPECT_NEAR(emd, emd0, 1e-6 * emd0);
+        EXPECT_MAT_NEAR(Mat(flow0), flow, 1e-6);
     }
+    // no cost and DIST_USER - error
+    {
+        Mat flow;
+        EXPECT_THROW(EMD(w1, w2, DIST_USER, noArray(), nullptr, flow), cv::Exception);
+        EXPECT_THROW(EMD(w1, w2, DIST_USER), cv::Exception);
+    }
+}
 
-    if( code < 0 )
-        ts->set_failed_test_info( code );
+TEST(Imgproc_EMD, distance_types)
+{
+    // 1D (sum = 210)
+    Matx<float, 4, 2> w1 {50, 1, 60, 2, 50, 3, 50, 4};
+    Matx<float, 5, 2> w2 {30, 1, 20, 2, 70, 3, 30, 4, 60, 5};
+
+    // 2D (sum = 210)
+    Matx<float, 4, 3> w3 {50, 0, 0, 60, 0, 1, 50, 1, 0, 50, 1, 1};
+    Matx<float, 5, 3> w4 {20, 0, 1, 70, 1, 0, 30, 1, 1, 60, 2, 2, 30, 3, 3};
+
+    // basic call with all distance types
+    {
+        const vector<DistanceTypes> good_types {DIST_L1, DIST_L2, DIST_C};
+        for (const auto& dt : good_types)
+        {
+            SCOPED_TRACE(cv::format("dt=%d", dt));
+            float emd = 0.f;
+            Mat flow;
+            // 1D
+            {
+                ASSERT_NO_THROW(emd = EMD(w1, w2, dt, noArray(), nullptr, flow));
+                const float emd0 = calcEMD(Mat(w1), Mat(w2), flow, dt, 1);
+                EXPECT_NEAR(emd0, emd, 1e-6);
+            }
+            // 2D
+            {
+                ASSERT_NO_THROW(emd = EMD(w3, w4, dt, noArray(), nullptr, flow));
+                const float emd0 = calcEMD(Mat(w3), Mat(w4), flow, dt, 2);
+                EXPECT_NEAR(emd0, emd, 1e-6);
+            }
+        }
+        const vector<DistanceTypes> bad_types {DIST_L12, DIST_FAIR, DIST_WELSCH, DIST_HUBER};
+        for (const auto& dt : bad_types)
+        {
+            SCOPED_TRACE(cv::format("dt=%d", dt));
+            EXPECT_THROW(EMD(w1, w2, dt), cv::Exception);
+        }
```

Comment:
It'll be dedicated test with parameter.

Reply:
I decided to extract this block to separate _invalid_ test. It seem to be too small to be in a separate parametrized test. `SCOPED_TRACE` would help with debugging test failures.